### PR TITLE
Support Invoke Transaction version 1

### PR DESCRIPTION
**Pull request by adriantpaez**
> ⚠️ The transaction definitions in the protobuf files change in this PR, so a previous database populated before this PR will be incompatible

## Description

StarkNet adds the new [Transaction Invoke version 1](https://docs.starknet.io/docs/Blocks/transactions#invoke-transaction-version-1), and we need to add the support to storage on our database.

## Changes:

- The new [`InvokeVersioned`](https://github.com/NethermindEth/juno/blob/adrian/invoke_txn_v1/internal/db/transaction/transaction.proto#L21) message is defined in the [`transaction.proto`](https://github.com/NethermindEth/juno/blob/adrian/invoke_txn_v1/internal/db/transaction/transaction.proto) file and represents an Invoke Transaction of any of the supported versions: [v0](https://docs.starknet.io/docs/Blocks/transactions#invoke-transaction-version-0) and [v1](https://docs.starknet.io/docs/Blocks/transactions#invoke-transaction-version-1)
- The new type [`TransactionInvokeV1`](https://github.com/NethermindEth/juno/blob/de7fe3d6fa6942136c03e24226856a9bf3aa5b8e/pkg/types/transaction.go#L36) is added to the `pkg/types` package and the old transaction type is renamed from `TransactionInvoke` to `TransactionInvokeV0`. The marshal.unmarshal process behind the [`GetTransaction`](https://github.com/NethermindEth/juno/blob/de7fe3d6fa6942136c03e24226856a9bf3aa5b8e/internal/db/transaction/manager.go#L41) and [`PutTransaction`](https://github.com/NethermindEth/juno/blob/de7fe3d6fa6942136c03e24226856a9bf3aa5b8e/internal/db/transaction/manager.go#L28) methods has been updated to support these new types.

## Types of changes

_Leave only items that describe your changes, remove the rest. Remove this line too._

- Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Testing

**Requires testing**: Yes

**Did you write tests??**: Yes

## Documentation

**If this requires a documentation update, did you add one?** Yes/No

### CODE REVIEW

**Review by stdevMac — APPROVED** [review]

File: internal/rpc/starknet/types.go [hunk 1]
```diff
@@ -172,13 +174,17 @@ func NewTxn(tx types.IsTransaction) (Txn, error) {
...
 	MaxFee    string   `json:"max_fee"`
 	Version   string   `json:"version"`
 	Signature []string `json:"signature"`
 	Nonce     string   `json:"nonce"`
-	Type      string   `json:"type"`
+}
+
+type CommonTxnProperties struct {
+	TxnHash string `json:"transaction_hash"`
+	BroadcastedTxnCommonProperties
```
Comment (tshakalekholoane) [inline internal/rpc/starknet/types.go:187]: Embedded fields should come first.

```suggestion
	BroadcastedTxnCommonProperties
	TxnHash string `json:"transaction_hash"`
```

File: internal/rpc/starknet/types.go [hunk 2]
```diff
@@ -218,7 +226,40 @@ func NewInvokeTxn(txn *types.TransactionInvoke) *InvokeTxn {
...
+
+func NewInvokeTxnV1(txn *types.TransactionInvokeV1) *InvokeTxnV1 {
+	signature := make([]string, len(txn.Signature))
+	for i, sig := range txn.Signature {
+		signature[i] = sig.Hex0x()
+	}
+	calldata := make([]string, len(txn.CallData))
+	for i, data := range txn.CallData {
+		calldata[i] = data.Hex0x()
+	}
```
Comment (tshakalekholoane) [inline internal/rpc/starknet/types.go:245]: We can avoid zeroing out the arrays.

File: internal/rpc/starknet/types.go [hunk 3]
```diff
@@ -247,29 +290,33 @@ func NewDeclareTxn(txn *types.TransactionDeclare) *DeclareTxn {
...
-	ContractAddress     string   `json:"contract_address"`
+	Type                TxnType  `json:"type"`
 	ContractAddressSalt string   `json:"contract_address_salt"`
 	ConstructorCalldata []string `json:"constructor_calldata"`
 }
 
+type DeployTxn struct {
+	TxnHash   string `json:"transaction_hash"`
+	ClassHash string `json:"class_hash"`
+	DeployTxnProperties
```
Comment (tshakalekholoane) [inline internal/rpc/starknet/types.go:303]: Embedded fields should appear first.

File: internal/sync/sync.go [hunk 4]
```diff
@@ -24,7 +24,11 @@ import (
 	"go.uber.org/zap"
 )
 
-var errIncompatibleStateRoot = errors.New("incompatible state root")
+var (
+	txnV0Id                  = new(felt.Felt).SetHex("0x0")
+	txnV1Id                  = new(felt.Felt).SetHex("0x1")
```
Comment (tshakalekholoane) [inline internal/sync/sync.go:29]: Assignment using the helpers `SetOne` and `SetZero` respectively are less costly.

**Review by tshakalekholoane — CHANGES_REQUESTED** [review]

**Review by tshakalekholoane — APPROVED** [review]

**Review by joshklop — APPROVED** [review]

**Pull request merged**